File: jira-mcp-server/tests/test_user_issues.rs
/// Integration tests for get_user_issues tool
mod common;

use common::{test_project_key, McpTestClient};
use serde_json::json;

#[test]
fn test_get_user_issues_default() {
    // Test get_user_issues with no parameters
    // Should return issues assigned to the current user
    // After bug fix #19, this should work correctly

    let mut client = McpTestClient::new().expect("Failed to create test client");

    let response = client
        .call_tool("get_user_issues", json!({}))
        .expect("Failed to call get_user_issues");

    let result =
        McpTestClient::extract_tool_result(&response).expect("Failed to extract tool result");

    // Verify response structure
    assert!(
        result.get("search_result").is_some(),
        "No search_result field"
    );
    assert!(
        result.get("resolved_user").is_some(),
        "No resolved_user field"
    );
    assert!(result.get("jql_query").is_some(), "No jql_query field");

    // Verify JQL does NOT have invalid syntax (bug #19)
    let jql = result["jql_query"]
        .as_str()
        .expect("jql_query is not a string");

    // ORDER BY should NOT be joined with AND
    assert!(
        !jql.contains(" AND ORDER BY"),
        "JQL should not have 'AND ORDER BY' (bug #19). Got: {}",
        jql
    );

    // Should have proper format: conditions ORDER BY ...
    assert!(
        jql.contains("ORDER BY updated DESC"),
        "JQL should end with 'ORDER BY updated DESC'. Got: {}",
        jql
    );

    // Verify user info
    let user_info = &result["resolved_user"];
    assert!(user_info.get("account_id").is_some(), "No account_id");
    assert!(user_info.get("display_name").is_some(), "No display_name");

    let is_current_user = user_info["is_current_user"]
        .as_bool()
        .expect("is_current_user is not a boolean");
    assert!(is_current_user, "User should be current user");

    // Verify search results
    let search_result = &result["search_result"];
    assert!(search_result.get("issues").is_some(), "No issues");
    assert!(search_result.get("total").is_some(), "No total");
}

#[test]
fn test_get_user_issues_with_status_filter() {
    // Test get_user_issues with status filter

    let mut client = McpTestClient::new().expect("Failed to create test client");

    let response = client
        .call_tool(
            "get_user_issues",
            json!({
                "status_filter": ["To Do", "In Progress"]
            }),
        )
        .expect("Failed to call get_user_issues");

    let result =
        McpTestClient::extract_tool_result(&response).expect("Failed to extract tool result");

    // Verify JQL syntax is correct (bug #19)
    let jql = result["jql_query"]
        .as_str()
        .expect("jql_query is not a string");
    assert!(!jql.contains(" AND ORDER BY"), "Invalid JQL syntax");
    assert!(jql.contains("ORDER BY updated DESC"), "Missing ORDER BY");

    // Verify status filter is applied
    assert!(
        jql.to_lowercase().contains("status"),
        "JQL should include status filter"
    );
}

#[test]
fn test_get_user_issues_with_project_filter() {
    // Test get_user_issues with project filter

    let mut client = McpTestClient::new().expect("Failed to create test client");
    let project_key = test_project_key();

    let response = client
        .call_tool(
            "get_user_issues",
            json!({
                "project_filter": [project_key.clone()]
            }),
        )
        .expect("Failed to call get_user_issues");

    let result =
        McpTestClient::extract_tool_result(&response).expect("Failed to extract tool result");

    // Verify JQL syntax (bug #19)
    let jql = result["jql_query"]
        .as_str()
        .expect("jql_query is not a string");
    assert!(!jql.contains(" AND ORDER BY"), "Invalid JQL syntax");
    assert!(jql.contains(&project_key), "JQL should include project key");

    // Verify all returned issues are from the specified project
    let issues = result["search_result"]["issues"]
        .as_array()
        .expect("issues is not an array");

    for issue in issues {
        let key = issue["key"].as_str().expect("Issue key is not a string");
        assert!(
            key.starts_with(&project_key),
            "Issue {} does not belong to project {}",
            key,
            project_key
        );
    }
}

#[test]
fn test_get_user_issues_with_issue_types() {
    // Test get_user_issues with issue type filter

    let mut client = McpTestClient::new().expect("Failed to create test client");

    let response = client
        .call_tool(
            "get_user_issues",
            json!({
                "issue_types": ["Task", "Story"]
            }),
        )
        .expect("Failed to call get_user_issues");

    let result =
        McpTestClient::extract_tool_result(&response).expect("Failed to extract tool result");

    // Verify JQL syntax (bug #19)
    let jql = result["jql_query"]
        .as_str()
        .expect("jql_query is not a string");
    assert!(!jql.contains(" AND ORDER BY"), "Invalid JQL syntax");
    assert!(
        jql.to_lowercase().contains("issuetype"),
        "JQL should include issue type filter"
    );
}

#[test]
fn test_get_user_issues_performance_metrics() {
    // Test that get_user_issues includes performance metrics

    let mut client = McpTestClient::new().expect("Failed to create test client");

    let response = client
        .call_tool("get_user_issues", json!({"limit": 5}))
        .expect("Failed to call get_user_issues");

    let result =
        McpTestClient::extract_tool_result(&response).expect("Failed to extract tool result");

    // Verify performance metrics
    assert!(
        result.get("performance").is_some(),
        "No performance metrics"
    );

    let perf = &result["performance"];
    assert!(perf.get("duration_ms").is_some(), "No duration_ms");
    assert!(perf.get("api_calls").is_some(), "No api_calls");
    assert!(perf.get("user_cache_hit").is_some(), "No user_cache_hit");
    assert!(
        perf.get("query_complexity").is_some(),
        "No query_complexity"
    );

    let duration = perf["duration_ms"]
        .as_u64()
        .or_else(|| perf["duration_ms"].as_i64().map(|v| v as u64))
        .expect("duration_ms is not a number");
    assert!(duration > 0, "Duration should be greater than 0");
}

#[test]
fn test_get_user_issues_combined_filters() {
    // Test get_user_issues with multiple filters combined

    let mut client = McpTestClient::new().expect("Failed to create test client");
    let project_key = test_project_key();

    let response = client
        .call_tool(
            "get_user_issues",
            json!({
                "project_filter": [project_key],
                "status_filter": ["To Do", "In Progress"],
                "issue_types": ["Task"],
                "limit": 5
            }),
        )
        .expect("Failed to call get_user_issues");

    let result =
        McpTestClient::extract_tool_result(&response).expect("Failed to extract tool result");

    // Verify JQL syntax with multiple filters (bug #19)
    let jql = result["jql_query"]
        .as_str()
        .expect("jql_query is not a string");

    assert!(
        !jql.contains(" AND ORDER BY"),
        "Invalid JQL syntax with multiple filters"
    );

    // All filters should be joined with AND, then ORDER BY at the end
    assert!(jql.contains("assignee"), "JQL should have assignee filter");
    assert!(
        jql.contains("ORDER BY updated DESC"),
        "JQL should end with ORDER BY"
    );

    // Multiple filters should be present
    let and_count = jql.matches(" AND ").count();
    assert!(
        and_count >= 2,
        "Expected at least 2 AND clauses for multiple filters"
    );
}

File: tests/test_user_issues.py
"""
Tests for get_user_issues tool.

Tests the user issue query functionality including:
- Get all user issues (default to current user)
- Get user issues with status filter
- Get user issues with project filter
- Get user issues with issue type filter
- Verify JQL syntax is correct (bug #19 fix)
"""

import pytest


@pytest.mark.asyncio
async def test_get_user_issues_default(mcp_session):
    """
    Test get_user_issues with no parameters.

    Should return issues assigned to the current user.
    After bug fix #19, this should work correctly.
    """
    result = await mcp_session.call_tool("get_user_issues", {})

    assert result is not None
    assert "content" in result

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify response structure
    assert "search_result" in data
    assert "resolved_user" in data
    assert "jql_query" in data

    # Verify JQL does NOT have invalid syntax (bug #19)
    jql = data["jql_query"]
    # ORDER BY should NOT be joined with AND
    assert " AND ORDER BY" not in jql
    # Should have proper format: conditions ORDER BY ...
    assert "ORDER BY updated DESC" in jql

    # Verify user info
    user_info = data["resolved_user"]
    assert "account_id" in user_info
    assert "display_name" in user_info
    assert user_info["is_current_user"] is True

    # Verify search results
    search_result = data["search_result"]
    assert "issues" in search_result
    assert "total" in search_result
    assert isinstance(search_result["issues"], list)


@pytest.mark.asyncio
async def test_get_user_issues_with_status_filter(mcp_session):
    """Test get_user_issues with status filter."""
    result = await mcp_session.call_tool(
        "get_user_issues", {"status_filter": ["To Do", "In Progress"]}
    )

    assert result is not None

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify JQL syntax is correct
    jql = data["jql_query"]
    assert " AND ORDER BY" not in jql
    assert "ORDER BY updated DESC" in jql

    # Verify status filter is applied
    assert "status" in jql.lower()

    # Verify applied filters
    assert "applied_filters" in data
    filters = data["applied_filters"]
    assert "status_categories" in filters or filters.get("status_categories") is not None


@pytest.mark.asyncio
async def test_get_user_issues_with_project_filter(mcp_session, test_project_key):
    """Test get_user_issues with project filter."""
    result = await mcp_session.call_tool(
        "get_user_issues", {"project_filter": [test_project_key]}
    )

    assert result is not None

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify JQL syntax
    jql = data["jql_query"]
    assert " AND ORDER BY" not in jql
    assert test_project_key in jql

    # Verify all returned issues are from the specified project
    issues = data["search_result"]["issues"]
    for issue in issues:
        assert issue["key"].startswith(test_project_key)


@pytest.mark.asyncio
async def test_get_user_issues_with_issue_types(mcp_session):
    """Test get_user_issues with issue type filter."""
    result = await mcp_session.call_tool(
        "get_user_issues", {"issue_types": ["Task", "Story"]}
    )

    assert result is not None

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify JQL syntax
    jql = data["jql_query"]
    assert " AND ORDER BY" not in jql
    assert "issuetype" in jql.lower()


@pytest.mark.asyncio
async def test_get_user_issues_with_priority_filter(mcp_session):
    """Test get_user_issues with priority filter."""
    result = await mcp_session.call_tool(
        "get_user_issues", {"priority_filter": ["High", "Critical"]}
    )

    assert result is not None

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify JQL syntax
    jql = data["jql_query"]
    assert " AND ORDER BY" not in jql
    assert "priority" in jql.lower()


@pytest.mark.asyncio
async def test_get_user_issues_with_due_date_filter(mcp_session):
    """Test get_user_issues with due date filter."""
    result = await mcp_session.call_tool(
        "get_user_issues", {"due_date_filter": "this_week"}
    )

    assert result is not None

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify JQL syntax
    jql = data["jql_query"]
    assert " AND ORDER BY" not in jql
    assert "due" in jql.lower()


@pytest.mark.asyncio
async def test_get_user_issues_pagination(mcp_session):
    """Test get_user_issues with pagination."""
    # Get first page
    result1 = await mcp_session.call_tool("get_user_issues", {"limit": 2, "start_at": 0})

    # Get second page
    result2 = await mcp_session.call_tool("get_user_issues", {"limit": 2, "start_at": 2})

    import json

    data1 = json.loads(result1["content"][0]["text"])
    data2 = json.loads(result2["content"][0]["text"])

    # Verify both pages have valid structure
    assert "search_result" in data1
    assert "search_result" in data2

    # If there are enough issues, pages should be different
    total = data1["search_result"]["total"]
    if total > 2:
        issues1 = {issue["key"] for issue in data1["search_result"]["issues"]}
        issues2 = {issue["key"] for issue in data2["search_result"]["issues"]}
        assert issues1 != issues2


@pytest.mark.asyncio
async def test_get_user_issues_combined_filters(mcp_session, test_project_key):
    """Test get_user_issues with multiple filters combined."""
    result = await mcp_session.call_tool(
        "get_user_issues",
        {
            "project_filter": [test_project_key],
            "status_filter": ["To Do", "In Progress"],
            "issue_types": ["Task"],
            "limit": 5,
        },
    )

    assert result is not None

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify JQL syntax with multiple filters
    jql = data["jql_query"]
    assert " AND ORDER BY" not in jql

    # All filters should be joined with AND, then ORDER BY at the end
    assert "assignee" in jql
    assert "ORDER BY updated DESC" in jql

    # Multiple filters should be present
    filter_count = jql.count(" AND ")
    assert filter_count >= 2  # At least project and status


@pytest.mark.asyncio
async def test_get_user_issues_performance_metrics(mcp_session):
    """Test that get_user_issues includes performance metrics."""
    result = await mcp_session.call_tool("get_user_issues", {"limit": 5})

    import json

    content = result["content"][0]["text"]
    data = json.loads(content)

    # Verify performance metrics
    assert "performance" in data
    perf = data["performance"]
    assert "duration_ms" in perf
    assert "api_calls" in perf
    assert "user_cache_hit" in perf
    assert "query_complexity" in perf
    assert isinstance(perf["duration_ms"], (int, float))
    assert perf["duration_ms"] > 0
